README.md
lightct
=======

[![Build Status](https://travis-ci.org/casimp/lightct.svg?branch=master)](https://travis-ci.org/casimp/lightct) [![Coverage Status](https://coveralls.io/repos/github/casimp/lightct/badge.svg?branch=master)](https://coveralls.io/github/casimp/lightct?branch=master)

What is lightct?
----------------

lightct is a python package for visual light computed tomography. With just a webcam and a rotating stage the package will allow for the acquisition of 2D projections and the reconstruction of objects in 3D.

This package will be showcased at the Royal Society Summer Exhibition (4th - 10th July 2016), allowing members of the public to run their own synchrotron style experiment:

https://royalsociety.org/events/summer-science-exhibition/exhibits/4d-science/

Requirements
------------

lightct is built on Python’s scientific stack (numpy, scipy, matplotlib, scikit-image). Development was carried out using the Anaconda (v 2.5.0) package manager, which is the recommended route for the installation of Python and the scientific stack:

https://www.continuum.io/downloads

lightct works with both Python 2 and 3, although Python 3 is now the recommended version unless you have a good reason to do otherwise! lightct has been tested under the following conditions:

-	Python: version 2.7, 3.4+
-	numpy: version 1.10
-	scipy: version 0.17
-	matplotlib: version 1.5
-	scikit-image: version 0.11

The package also leverages OpenCV for the acquisition of images:

-	opencv: version 3.1

OpenCV can be installed on an Anaconda flavoured build of Python by running the following in the command prompt:

```
conda install -c https://conda.binstar.org/menpo opencv3
```

Installation of lightct
-----------------------

Installing lightct is easily done using pip. Assuming you have pip installed (included with the Anaconda build), just run the following from the command-line:

```
pip install lightct
```

This command will download the stable version of lightct from the Python Package Index and install it to your system.

Alternatively, you can install the most recent distribution using the setup.py script. The source is stored in the GitHub repo, which can be browsed at:

https://github.com/casimp/lightct

Simply download and unpack, then navigate to the download directory and run the following from the command-line:

```
python setup.py install
```

Documentation
-------------

Please see the lightct-guide for more information:

https://github.com/casimp/lightct/blob/master/docs/lightct_guide.ipynb


CHANGES.txt
v 0.1.0, 20160527 -- First stable release
 - acquires projections, recentres, crops and reconstructs.

lightct/tests/load_tests.py
# -*- coding: utf-8 -*-
"""
Created on Thu May 26 22:25:14 2016

@author: casim
"""

from nose.tools import assert_equal
from lightct.load_scan import LoadProjections
import os
from mock import patch


proj_path = os.path.join(os.path.dirname(os.path.abspath(__file__)), r'test_scan/projections')

def test_load():
    """
    Check that the correct number of files is being loaded.
    """
    scan_data = LoadProjections(proj_path)
    num_tifs = len([i for i in os.listdir(proj_path) if i[-4:] == '.tif'])
    assert_equal(scan_data.im_stack.shape[-1], num_tifs)


@patch("matplotlib.pyplot.show")
def test_auto_angles(mock_show):
    mock_show.return_value = None
    scan_data = LoadProjections(proj_path)
    scan_data.auto_set_angles(20)
    
@patch("matplotlib.pyplot.show")
def test_plot_hist(mock_show):
    mock_show.return_value = None
    scan_data = LoadProjections(proj_path)
    scan_data.plot_histogram()
    
    
@patch("matplotlib.pyplot.show")
def test_manual_angles(mock_show):
    mock_show.return_value = None
    scan_data = LoadProjections(proj_path)
    scan_data.manual_set_angles()
    

def test_set_angles():
    scan_data = LoadProjections(proj_path)
    scan_data.set_angles(21, 360)
    
    
@patch("matplotlib.pyplot.show")
def test_auto_centre(mock_show):
    mock_show.return_value = None
    scan_data = LoadProjections(proj_path)
    scan_data.set_angles(21, 360)
    scan_data.auto_centre()
    
def test_set_centre():
    scan_data = LoadProjections(proj_path)
    scan_data.set_angles(21, 360)
    scan_data.set_centre(-40)
    assert_equal(scan_data.cor_offset, -40)


@patch("matplotlib.pyplot.show")
def test_set_crop(mock_show):
    mock_show.return_value = None
    scan_data = LoadProjections(proj_path)
    scan_data.set_angles(21, 360)
    scan_data.set_centre(-40)
    
    assert_equal(scan_data.crop, (None, None, None, None))
    scan_data.set_crop(100, 100, 100)
    assert_equal(scan_data.crop, (100, -100, 100, -100))
    scan_data.set_crop(0, 0, 0, plot=False)
    assert_equal(scan_data.crop, (None, None, None, None))
    
    
def test_reconstruction():
    scan_data = LoadProjections(proj_path)
    scan_data.set_angles(21, 360)
    scan_data.set_centre(-40)
    scan_data.set_crop(300, 100, 100, plot=False)
    scan_data.reconstruct(downsample=(4,4), crop=True, 
                          median_filter=True, kernel=3, save=False)
    scan_data.reconstruct(downsample=(8,8), crop=False, 
                          median_filter=False, save=True)
    scan_data.reconstruct(downsample=(16,16), recon_alg='sart', sart_iters=2,
                          save=False)


lightct/tests/scan_tests.py
# -*- coding: utf-8 -*-
"""
Created on Thu May 26 22:25:14 2016

@author: casim
"""

from lightct.tomo_scan import TomoScan
from mock import patch
import numpy as np

@patch("lightct.tomo_scan.image_acquisition")
def testfunction(mfun):
    mfun.return_value  = np.random.rand(20, 10, 3)
    TomoScan(10, '', save=False)


lightct/tomo_lego.py
from __future__ import print_function, division

from mayavi import mlab

import numpy as np
import scipy as sc
import matplotlib.pyplot as plt
from skimage.io import imread
from skimage import data_dir
from skimage.transform import radon, rescale
from skimage.transform import iradon
from PIL import Image
from skimage.transform import iradon_sart
import cv2
import time
import skimage
from skimage import color
from skimage import measure
from skimage.filters import threshold_otsu
from skimage import img_as_ubyte


PREFIX = 'snowman_'
nproj = 20
nstep = nproj+1

print(nstep,nproj)

angle = np.linspace(0,360,nstep*2,dtype=int)


camera_port = 0
camera = cv2.VideoCapture(0)
start = time.clock()
for i in range (20) :
    calibi=angle[i]
    retval,im = camera.read()
    im = (skimage.color.rgb2hsv(im))[:,:,2]
end=time.clock()
del(camera)
width = im.shape[1]
height = im.shape[0]
overhead = (end-start)/50



time_half_rotation = 3.2
#time_half_rotation = int(raw_input('Time for 180 rotation (s) : '))


tproj = time_half_rotation/nproj

if tproj<(overhead*1):
    raise Exception('Too many projections')


projection=np.zeros((height,width,nstep*2))

camera = cv2.VideoCapture(camera_port)
for i in range (nstep*2) :
    temp = angle[i]
    retval,imtmp=camera.read()
    projection[:,:,i] = (skimage.color.rgb2hsv(imtmp))[:,:,2]
#    projection[:,:,i,0]=imtmp[:,:,0]
#    projection[:,:,i,1]=imtmp[:,:,1]
#    projection[:,:,i,2]=imtmp[:,:,2]
    time.sleep((tproj-overhead))
del(camera)

downsample = skimage.transform.downscale_local_mean(projection,(4,4,1))
for i in range (nstep*2) :
    downsample[:,:,i] = sc.signal.medfilt(downsample[:,:,i],kernel_size=9)

down_height = downsample.shape[0]
down_width = downsample.shape[1]

data=np.zeros((down_width,down_width,down_height))
for j in range (0,down_height) :
    sinotmp = np.squeeze(downsample[j,:,:])
    imagetmp = iradon(sinotmp,theta=angle,filter=None,circle=True)
   # imagetmp = iradon_sart(sinotmp, theta = angle)
    #imagetmp2=iradon_sart(sinotmp, theta = angle,image=imagetmp)
    #imagetmp3=iradon_sart(sinotmp, theta = angle,image=imagetmp2)
    #imagetmp4=iradon_sart(sinotmp, theta = angle,image=imagetmp3)
    sc.misc.imsave('/Users/lcourtois/Documents/Perso/3Dmagination/Software/TomoLego/FBP_'+'%04d' % j+'.tif',imagetmp)
    data[:,:,j]=imagetmp
    
    
    
    
    
data = data[60:down_width-60,60:down_width-60,:]
datafilter=skimage.transform.downscale_local_mean(data,(2,2,2))
for i in range (0,datafilter.shape[2]) :
    datafilter[:,:,i]=sc.signal.medfilt(datafilter[:,:,i],kernel_size=9)



datathres=np.zeros((datafilter.shape[0],datafilter.shape[1],datafilter.shape[2]))
for i in range (0,datafilter.shape[2]) :
    tempotsu=threshold_otsu(datafilter[:,:,i])+0.08
    if ((np.max(datafilter[:,:,i]))>=0.35) :
        datathres[:,:,i]=(datafilter[:,:,i]>=tempotsu)
    else :
        datathres[:,:,i]=np.zeros((datafilter.shape[0],datafilter.shape[1]))
#datatemp = sc.ndimage.morphology.binary_opening(datathres, iterations=7)
#datatemp = sc.ndimage.morphology.binary_closing(datatemp, iterations=5)
verts, faces = measure.marching_cubes(datathres, 0)
mlab.triangular_mesh([vert[0] for vert in verts],[vert[1] for vert in verts],[vert[2] for vert in verts],faces)
mlab.show()

#datatemp=datatemp *10
#src = mlab.pipeline.scalar_field(datatemp)
#mlab.pipeline.iso_surface(src, contours=5, opacity=1)



lightct/tomo_scan.py
# -*- coding: utf-8 -*-
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

import sys
import time
import os

import cv2
from scipy.misc import imsave
import numpy as np
from skimage import color

from lightct.load_scan import LoadProjections

def image_acquisition(num_proj, camera_port=0, wait=0, hsv='v'):
    hsv_dict = {'h':0, 's':1, 'v':2}
    camera = cv2.VideoCapture(camera_port)
    camera.set(3, 2000)
    camera.set(4, 2000)
    try:
        dims = camera.read()[1][:, :, 2].shape + (num_proj, )
    except TypeError:
        error = ('Camera returning None. Check camera settings (port) and'
                 ' ensure camera is not being run by other software.')
        raise TypeError(error)
    im_stack = np.zeros(dims)

    # Acquires defined number of images (saves slice from hsv)
    for i in range(num_proj):
        _, im = camera.read()
        im_stack[:, :, i] = color.rgb2hsv(im)[:, :, hsv_dict[hsv]]
        sys.stdout.write('\rProgress: [{0:20s}] {1:.0f}%'.format('#' *
                         int(20*(i + 1) / num_proj),
                         100*((i + 1)/num_proj)))
        sys.stdout.flush()
        time.sleep(wait)
    del camera
    return im_stack


class TomoScan(LoadProjections):
    
    def __init__(self, num_proj, folder, camera_port=0,
                 wait=0, save=True, hsv='v'):
        """
        Acquires specified number of projections for analysis and
        reconstruction. The specified number must ensure that a rotational
        range > 360 degrees is captured.

        # num_proj:  Number of projections to acquire (must cover > 360deg)
        # folder:    Path to folder for data storage. Projections and
                     reconstructed slices will be saved here.
        # hsv:       Extract either the hue (h), saturation (s) or variance(v)
                     from the image matrix.
        """
        self.folder = folder
        self.p0 = 0
        self.cor_offset = 0
        self.crop = None, None, None, None
        self.num_images = None
        self.angles = None
        self.recon_data = None
        
        self.im_stack = image_acquisition(num_proj, camera_port, wait, hsv)
        self.height, self.width = self.im_stack.shape[:2]

        if save:
            save_folder = os.path.join(self.folder, 'projections')
            if not os.path.exists(save_folder):
                os.makedirs(save_folder)
            for idx in range(self.im_stack.shape[-1]):
                f_path = os.path.join(save_folder, '%04d.tif' % idx)
                imsave(f_path, self.im_stack[:, :, idx])

lightct/plot_funcs.py
import matplotlib
from matplotlib import patches
import matplotlib.pyplot as plt
from matplotlib.widgets import Slider, Button
import numpy as np


def recentre_plot(image, cor_offset):
    fig, ax_array = plt.subplots(1, 2, figsize=(10, 6))
    height, width = image.shape
    if cor_offset <= 0:
        poly_pnts = [[width + cor_offset, 0], [width, 0],
                     [width, height], [width + cor_offset, height]]
    else:
        poly_pnts = [[0, 0], [cor_offset, 0],
                     [cor_offset, height], [0, height]]

    ax_array[0].imshow(image)
    centre = width / 2 + cor_offset / 2
    ax_array[0].plot([centre, centre], [0, height], 'k-',
                     linewidth=2, label='New COR')
    ax_array[0].plot([width / 2, width / 2],
                     [0, height], 'r-', linewidth=2, label='Old COR')
    ax_array[0].legend()
    ax_array[0].set_xlim([0, image.shape[1]])
    ax_array[0].set_ylim([image.shape[0], 0])

    if cor_offset >= 0:
        offset_image = np.copy(image[:, cor_offset:])
    else:
        offset_image = np.copy(image[:, :cor_offset])

    ax_array[1].imshow(offset_image)
    ax_array[1].plot([offset_image.shape[1] / 2, offset_image.shape[1] / 2],
                     [0, height], 'k-', linewidth=2, label='New COR')
    ax_array[1].legend()

    ax_array[1].set_xlim([0, offset_image.shape[1]])
    ax_array[1].set_ylim([offset_image.shape[0], 0])

    ax_array[0].add_patch(patches.Polygon(poly_pnts, closed=True,
                                          fill=False, hatch='///', color='k'))
    ax_array[0].set_title('Uncropped')
    ax_array[1].set_title('Cropped and centred')

    plt.show()


class SetAngleInteract(object):

    def __init__(self, im_stack, p0):
        self.im_stack = im_stack
        self.p0 = p0
        self.num_images = None
        self.angles = None

    def interact(self):
        backend = matplotlib.get_backend()
        err = ("Matplotlib running inline. Plot interaction not possible."
               "\nTry running %matplotlib in the ipython console (and "
               "%matplotlib inline to return to default behaviour). In "
               "standard console use matplotlib.use('TkAgg') to interact.")

        assert backend != 'module://ipykernel.pylab.backend_inline', err
        fig, ax_array = plt.subplots(1, 2, figsize=(10, 5))

        ax_slider = plt.axes([0.2, 0.07, 0.5, 0.05])
        ax_button = plt.axes([0.81, 0.05, 0.1, 0.075])

        for ax in ax_array:
            ax.imshow(self.im_stack[:, :, self.p0])
            ax.set_xticklabels([])
            ax.set_yticklabels([])
        fig.subplots_adjust(bottom=0.2)
        nfiles = self.im_stack.shape[-1] + 1
        window_slider = Slider(ax_slider, 'Image', self.p0, nfiles, valinit=0)
        store_button = Button(ax_button, r'Save - 360')

        def slider_update(val):
            ax_array[1].imshow(self.im_stack[:, :, int(window_slider.val)])
            window_slider.valtext.set_text('%i' % window_slider.val)
            fig.canvas.draw_idle()

        window_slider.on_changed(slider_update)

        def store_data(label):
            # Check this is correct - proj_ref!!!
            self.num_images = int(window_slider.val) - self.p0 + 1
            self.angles = np.linspace(0, 360, self.num_images)
            plt.close()

        store_button.on_clicked(store_data)
        plt.show()
        return window_slider, store_button


lightct/tomo_lego_rev.py
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

from mayavi import mlab

import os
import numpy as np
import scipy as sc
import matplotlib.pyplot as plt
from skimage.io import imread
from skimage import data_dir
from skimage.transform import radon, rescale
from skimage.transform import iradon
from PIL import Image
from skimage.transform import iradon_sart
import cv2
import time
import skimage
from skimage import color
from skimage import measure
from skimage.filters import threshold_otsu
from skimage import img_as_ubyte

class TomoScan(object):
    
    def __init__(self, camera_port = 0, nproj = 20, ang_range = [0, 360]):
        
        self.camera_port = camera_port
        self.nproj = nproj
        self.nstep = nproj + 1
        self.angle = np.linspace(ang_range[0], ang_range[1], self.nstep * 2, dtype = int)
        print(self.nstep, self.nproj)
        
    def image_timings(self):


        camera = cv2.VideoCapture(self.camera_port)
        start = time.clock()
        for i in range (20) :
            retval,im = camera.read()
            im = (skimage.color.rgb2hsv(im))[:,:,2]
        end=time.clock()
        del(camera)
        self.width = im.shape[1]
        self.height = im.shape[0]
        self.overhead = (end-start) / 50

        #time_half_rotation = int(raw_input('Time for 180 rotation (s) : '))
        time_half_rotation = 3.2 ######!!!!!
        self.tproj = time_half_rotation / self.nproj
        
        if self.tproj<(self.overhead * 1):
            raise Exception('Too many projections')
        
    def scan(self, reconstruct = True, save = True):
        
        self.projection=np.zeros((self.height, self.width, self.nstep*2))
        camera = cv2.VideoCapture(self.camera_port)
        for i in range (self.nstep * 2) :
            #temp = self.angle[i]
            retval, imtmp = camera.read()
            self.projection[:, :, i] = (skimage.color.rgb2hsv(imtmp))[:, :, 2]

            time.sleep((self.tproj - self.overhead))
        del(camera)
        
        if reconstruct:
            self.reconstruct()
    
    def load_projs(self, folder):
        
        files =  [file for file in os.listdir(folder) if file[-4:] == '.tif']
        self.im_stack = np.zeros((sc.misc.imread(folder + files[0]).shape + (len(files), )))
            
        for idx, file in enumerate(files):
            self.im_stack[:, :, idx] = sc.misc.imread(folder + file) 
        
    def reconstruct(self, downsample = (4, 4, 1), pre_filter = 'median', kernel = 9, save_folder = ''):
        
        downsample = skimage.transform.downscale_local_mean(self.projection, downsample)
        
        if pre_filter != False:
            pre_filter = sc.signal.medfilt if pre_filter == 'median' else pre_filter
            for i in range (self.nstep * 2): 
                downsample[:, :, i] = sc.signal.medfilt(downsample[:, :, i], kernel_size = kernel)
        
        down_height = downsample.shape[0]
        down_width = downsample.shape[1]
        
        self.data = np.zeros((down_width,down_width,down_height))
        
        for j in range (0, down_height) :
            sinotmp = np.squeeze(downsample[j, :, :])
            imagetmp = iradon(sinotmp, theta = self.angle, 
                              filter = None, circle = True)

            sc.misc.imsave('/Users/lcourtois/Documents/Perso/3Dmagination/Software/TomoLego/FBP_'+'%04d' % j+'.tif',imagetmp)
            self.data[:, :, j] = imagetmp
    
    def vizualize(self, crop = 60, downsample = (2, 2, 2), filter = True, kernel = 9):
        
        self.data = self.data[crop: -crop, crop: -crop, :]
        
        datafilter = skimage.transform.downscale_local_mean(self.data, downsample)
        if filter:
            for i in range(datafilter.shape[2]):
                datafilter[:, :, i]=sc.signal.medfilt(datafilter[:, :, i], kernel_size = kernel)
        
        
        
        datathres=np.zeros((datafilter.shape[0],datafilter.shape[1],datafilter.shape[2]))
        for i in range (0,datafilter.shape[2]) :
            tempotsu=threshold_otsu(datafilter[:, :, i]) + 0.08
            if (np.max(datafilter[:, :, i])) >= 0.35:
                datathres[:, :, i]=(datafilter[:, :, i] >= tempotsu)
            else :
                datathres[:,:,i] = np.zeros((datafilter.shape[0], datafilter.shape[1]))

        verts, faces = measure.marching_cubes(datathres, 0)
        mlab.triangular_mesh([vert[0] for vert in verts],[vert[1] for vert in verts],[vert[2] for vert in verts],faces)
        mlab.show()
        
        #datatemp=datatemp *10
        #src = mlab.pipeline.scalar_field(datatemp)
        #mlab.pipeline.iso_surface(src, contours=5, opacity=1)



lightct/load_scan.py
# -*- coding: utf-8 -*-
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

import sys
import os

import matplotlib.pyplot as plt
import numpy as np
from scipy.misc import imread, imsave
from scipy.signal import medfilt, argrelmin
from skimage.transform import iradon, iradon_sart, downscale_local_mean

from lightct.plot_funcs import recentre_plot, SetAngleInteract


class LoadProjections(object):

    def __init__(self, folder, im_type='tif'):
        """
        Load a previously acquired series of projections for analysis and
        reconstruction.

        # folder:    Path to folder where projections are stored. Reconstructed
                     slices will also be saved here.
        """
        self.folder = folder
        self.p0 = 0
        self.cor_offset = 0
        self.crop = None, None, None, None
        self.num_images = None
        self.angles = None

        files = [f for f in os.listdir(folder) if f[-4:] == '.%s' % im_type]
        im_shape = imread(os.path.join(self.folder, files[0])).shape
        self.im_stack = np.zeros(im_shape + (len(files), ))
        for idx, fname in enumerate(files):
                sys.stdout.write('\rProgress: [{0:20s}] {1:.0f}%'.format('#' *
                                 int(20*(idx + 1) / len(files)),
                                 100*((idx + 1)/len(files))))
                sys.stdout.flush()
                f = os.path.join(self.folder, fname)
                self.im_stack[:, :, idx] = imread(f)

        self.height = self.im_stack.shape[0]
        self.width = self.im_stack.shape[1]
        
    def plot_histogram(self, proj=5):
        """
        Plots histogram of pixel intensity for specified projection.
        
        # proj:       Number of projection to display/assess (int)
        """
        histogram = np.histogram(self.im_stack[:, :, proj], 255)
        plt.plot(histogram[0])
        plt.show()

    def set_angles(self, num_images, ang_range=360, p0=5):
        """
        Manually define the number of images in 360 degrees. 

        # num_images: Specify number of images
        # ang_range:  Specify angular range (must be multiple of 180)
        # p0:         Projection to use as initial or reference projection.
                      Recommended to be greater than 1 (due to acquisition
                      spacing issues in initial projections)

        """
        self.p0 = p0

        error = 'Images must cover a rotational range of 180 or 360 deg'
        assert (ang_range == 180) or (ang_range == 360), error
        self.angles = np.linspace(0, ang_range, num_images)
        self.num_images = num_images
        
    def auto_set_angles(self, est_nproj, p0=5, downscale=True, plot=True):
        """
        Attempts to automatically locate image at 360 degrees (and multiples
        of 360 degrees). Alignment based on difference calculation between 
        reference projection each subsequent projections. 
        
        # est_nproj:  Estimated number of projections in 360 degrees
        # p0:         Projection to use as initial or reference projection.
                      Recommended to be greater than 1 (due to acquisition
                      spacing issues in initial projections)
        # plot:       Plot the difference results
        """
        order = est_nproj // 2
        self.p0 = p0
        
        diff = np.nan * np.ones((self.im_stack.shape[-1] - p0))
        proj_nums = range(p0, self.im_stack.shape[-1])

        ref = self.im_stack[:, :, p0]
        data = self.im_stack[:, :, p0:]
        
        tmp = data - ref[:, :, np.newaxis]
        tmp = tmp.reshape(-1, tmp.shape[-1])
    
        diff = tmp.std(axis=0)

        minimas = argrelmin(diff, order=order)
        self.num_images = minimas[0][0] + 1
        self.angles = np.linspace(0, 360, self.num_images, dtype=int)
        
        if plot:
            plt.figure()
            plt.plot(proj_nums, diff)
            plt.plot(minimas[0] + p0, np.array(diff)[minimas], 'r*')
            plt.plot([minimas[0][0] + p0, minimas[0][0] + p0],
                     [0, np.max(diff)], 'r--')
            plt.xlabel('Image number')
            plt.ylabel('Thresholded Pixels Relative to Image 1')
            plt.text(minimas[0][0] + p0, np.max(diff), r'$360^{\circ}$',
                     horizontalalignment='center', verticalalignment='bottom')
        plt.show()
                     
        print('\n%i images in a 360 rotation.\n\nIf this is incorrect '
              'either rerun with a different value for est_nproj or use the '
              'manual method.' % self.num_images)
        
    def set_centre(self, cor):
        """
        Define the centre of rotation.
        
        # window:     Window width to search across (pixels).
        """
        self.cor_offset = cor
        
    def auto_centre(self, window=400, downsample_y=2, plot=True):
        """
        Automatic method for finding the centre of rotation.
        
        # window:     Window width to search across (pixels).
        """
        downsample = (downsample_y, 1)
        half_win = window // 2
        win_range = range(-half_win, half_win)
        
        # Compare ref image with flipped 180deg counterpart
        ref_image = self.im_stack[:, half_win:-half_win, self.p0]
        ref = downscale_local_mean(ref_image, downsample)
        im_180_image = self.im_stack[:, :, self.num_images // 2 + self.p0]
        im_180 = downscale_local_mean(im_180_image, downsample)
        flipped = np.fliplr(im_180)
        
        diff = np.nan * np.zeros(len(win_range))

        # Flip win_range as we are working on flipped data
        for idx, i in enumerate(win_range[::-1]):
            
            cropped = flipped[:, half_win + i: -half_win + i]
            tmp = cropped - ref
            diff[idx] = tmp.std()
            
        minima = np.argmin(diff)
        self.cor_offset = win_range[minima]
        print('COR = %i' % self.cor_offset)

        if plot:
            plt.plot([i for i in win_range], diff)
            plt.plot(self.cor_offset, np.min(diff), '*')
            plt.ylabel('Standard deviation (original v 180deg flipped)')
            plt.xlabel('2 * Centre correction (pixels)')
            im_copy = np.copy(self.im_stack[:, :, self.p0])
            recentre_plot(im_copy, self.cor_offset)
        
    def manual_set_angles(self, p0=5):
        """
        Manually define the number of images in 360 degrees. Defaults to 
        interactive mode in which images can be compared against initial, 
        reference image.

        # p0:         Projection to use as initial or reference projection.
                      Recommended to be greater than 1 (due to acquisition
                      spacing issues in initial projections)
        """
        self.p0 = p0

        interact = SetAngleInteract(self.im_stack, self.p0)
        interact.interact()
        self.angles = interact.angles
        self.num_images = interact.num_images
        
    def set_crop(self, width, top, bottom, plot=True):
        """
        Crop...
        """
        if self.cor_offset >= 0:
            images = self.im_stack[:, self.cor_offset:]
        else:
            images = self.im_stack[:, :self.cor_offset]
            
        self.crop = ()
        for i in (width, -width, top, -bottom): 
            self.crop += (None,) if i == 0 else (i,)

        
        if plot:
            left, right, top, bottom = self.crop
            
            images_per_degree = self.num_images / 360
            degrees = [0, 60, 120]
            image_nums = [int(images_per_degree * deg) for deg in degrees]
            fig, ax_array = plt.subplots(1, 3, figsize=(8, 3))
            for ax, i in zip(ax_array, image_nums):
                ax.imshow(images[top:bottom, left:right, i])
                ax.text(0.15, 0.88, r'$%0d^\circ$' % (i * 360/self.num_images), 
                        size=14, transform = ax.transAxes,
                        va = 'center', ha = 'center')
                ax.xaxis.set_ticklabels([])
                ax.yaxis.set_ticklabels([])
            fig.tight_layout()
            plt.show()
                
    def reconstruct(self, downsample=(4, 4), crop=True, median_filter=False, 
                    kernel=9, recon_alg='fbp', sart_iters=1, 
                    crop_circle=True, save=True):
        """
        Reconstruct the data using a radon transform. Reconstructed slices
        saved in folder specified upon class creation.

        # downsample: Downsample (local mean) data before reconstructing.
                      Specify mean kernel size (height, width).
        # pre_filter: If True apply median filter to data before reconstructing
        # kernel:     Kernel size to use for median filter
        """
            
        if self.cor_offset >= 0:
            images = self.im_stack[:, self.cor_offset:]
        else:
            images = self.im_stack[:, :self.cor_offset]
            
        images = images[:, :, self.p0:self.num_images + self.p0]
        
        if crop:
            left, right, top, bottom = self.crop
            images = images[top:bottom, left:right]
            
        images = downscale_local_mean(images, downsample + (1, ))
        recon_height, recon_width = images.shape[:2]
        self.recon_data = np.zeros((recon_width, recon_width, recon_height))
        
#        if crop_circle:
#            w = int((recon_width**2 / 2)**0.5) 
#            w = w if (w - recon_width) % 2 == 0 else w - 1
#            w0 =   int((recon_width - w) / 2  )
#            wf = int(w0 + w)
#            self.recon_data = self.recon_data[w0:wf, w0:wf]

        if median_filter:
            print('Applying median filter...')
            for i in range(images.shape[-1]):
                sys.stdout.write('\rProgress: [{0:20s}] {1:.0f}%'.format('#' *
                                 int(20 * (i + 1) / images.shape[-1]),
                                 100 * ((i + 1) / images.shape[-1])))
                sys.stdout.flush()
                images[:, :, i] = medfilt(images[:, :, i], kernel_size=kernel)

        print('\nReconstructing...')
        for j in range(recon_height):
            sys.stdout.write('\rProgress: [{0:20s}] {1:.0f}%'.format('#' *
                             int(20 * (j + 1) / recon_height),
                             100 * ((j + 1) / recon_height)))
            sys.stdout.flush()
            sino_tmp = np.squeeze(images[j, :, :])
            
            if recon_alg is 'sart':
                image_tmp = iradon_sart(sino_tmp, theta=self.angles)
                for i in range(sart_iters - 1):
                    image_tmp = iradon_sart(sino_tmp, theta=self.angles, 
                                            image=image_tmp)
            else:
                image_tmp = iradon(sino_tmp, theta=self.angles, 
                                   filter=None, circle=True)
#            if crop_circle:
#                image_tmp = image_tmp[w0:wf, w0:wf]
                
            self.recon_data[:, :, j] = image_tmp
            if save:            
                save_folder = os.path.join(self.folder, 'reconstruction')
                if not os.path.exists(save_folder):
                    os.makedirs(save_folder)
                imsave(os.path.join(save_folder, '%04d.tif' % j), image_tmp)

